// @triton-one/yellowstone-grpc/dist/esm/index.js
/** TypeScript/JavaScript client for gRPC Geyser. */
// Import generated gRPC client and types.
import { SubscribeUpdateTransactionInfo } from "./grpc/geyser.js";
// Reexport automatically generated types
export { CommitmentLevel, SubscribeRequest, SubscribeRequest_AccountsEntry, SubscribeRequest_BlocksEntry, SubscribeRequest_BlocksMetaEntry, SubscribeRequest_SlotsEntry, SubscribeRequest_TransactionsEntry, SubscribeRequestAccountsDataSlice, SubscribeRequestFilterAccounts, SubscribeRequestFilterAccountsFilter, SubscribeRequestFilterAccountsFilterLamports, SubscribeRequestFilterAccountsFilterMemcmp, SubscribeRequestFilterBlocks, SubscribeRequestFilterBlocksMeta, SubscribeRequestFilterEntry, SubscribeRequestFilterSlots, SubscribeRequestFilterTransactions, SubscribeUpdate, SubscribeUpdateAccount, SubscribeUpdateAccountInfo, SubscribeUpdateBlock, SubscribeUpdateBlockMeta, SubscribeUpdatePing, SubscribeUpdateSlot, SubscribeUpdateTransaction, SubscribeUpdateTransactionInfo, } from "./grpc/geyser.js";
import { Duplex } from "stream";
import * as napi from "./napi/index.js";
export default class Client {
    _insecureEndpoint;
    _insecureXToken;
    _channelOptions;
    _grpcClient = null;
    constructor(endpoint, xToken, channel_options) {
        this._insecureEndpoint = endpoint;
        this._insecureXToken = xToken;
        this._channelOptions = channel_options;
    }
    async connect() {
        // Use the factory method to create the client
        this._grpcClient = await napi.GrpcClient.new(this._insecureEndpoint, this._insecureXToken, this._channelOptions);
    }
    async getLatestBlockhash(commitment) {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        const request = {
            commitment: commitment ?? null,
        };
        return await this._grpcClient.getLatestBlockhash(request);
    }
    async ping(count) {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        const request = {
            count,
        };
        return (await this._grpcClient.ping(request)).count;
    }
    async getBlockHeight(commitment) {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        const request = {
            commitment,
        };
        return (await this._grpcClient.getBlockHeight(request)).blockHeight;
    }
    async getSlot(commitment) {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        const request = {
            commitment,
        };
        return (await this._grpcClient.getSlot(request)).slot;
    }
    async isBlockhashValid(blockhash, commitment) {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        const request = {
            blockhash,
            commitment,
        };
        return await this._grpcClient.isBlockhashValid(request);
    }
    async getVersion() {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        const request = {};
        return (await this._grpcClient.getVersion(request)).version;
    }
    async subscribeReplayInfo() {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        const request = {};
        return await this._grpcClient.subscribeReplayInfo(request);
    }
    async subscribe() {
        // Inner stream.Duplex config passed to both stream.Readable and Writable.
        // See: https://nodejs.org/en/blog/feature/streams2#new-streamduplexoptions
        const options = {
            // Pass objects not bytes.
            objectMode: true,
            // Skip unnecessary buffer conversion for performance.
            decodeStrings: false,
            // TODO: Fine tune high watermark for performance and backpressure.
            // highWaterMark: 16
        };
        const stream = this._grpcClient.subscribe();
        return new Promise((resolve, reject) => {
            try {
                resolve(new ClientDuplexStream(stream, options));
            }
            catch (err) {
                reject(err);
            }
        });
    }
}
class ClientDuplexStream extends Duplex {
    _napiDuplexStream;
    constructor(stream, options) {
        super({ ...options });
        this._napiDuplexStream = stream;
    }
    async _read(_size) {
        try {
            const update = await this._napiDuplexStream.read();
            this.push(update);
        }
        catch (err) {
            this.push(null); // Signal end of stream
            this.destroy(err); // Handle resource cleanup
        }
    }
    _write(chunk, _encoding, callback) {
        try {
            this._napiDuplexStream.write(chunk);
            callback();
        }
        catch (err) {
            callback(err);
        }
    }
}
export const txEncode = {
    encoding: napi.WasmUiTransactionEncoding,
    encode_raw: napi.encodeTx,
    encode: (message, encoding, max_supported_transaction_version, show_rewards) => {
        return JSON.parse(napi.encodeTx(SubscribeUpdateTransactionInfo.encode(message).finish(), encoding, max_supported_transaction_version, show_rewards));
    },
};
export const txErrDecode = {
    decode_raw: napi.decodeTxError,
    decode: (buf) => {
        return JSON.parse(napi.decodeTxError(Array.from(buf)));
    },
};
