// @triton-one/yellowstone-grpc/dist/cjs/napi/index.js
var readFileSync = require('node:fs').readFileSync;
var nativeBinding = null;
var loadErrors = [];
var isMusl = function () {
    var musl = false;
    if (process.platform === 'linux') {
        musl = isMuslFromFilesystem();
        if (musl === null) {
            musl = isMuslFromReport();
        }
        if (musl === null) {
            musl = isMuslFromChildProcess();
        }
    }
    return musl;
};
var isFileMusl = function (f) { return f.includes('libc.musl-') || f.includes('ld-musl-'); };
var isMuslFromFilesystem = function () {
    try {
        return readFileSync('/usr/bin/ldd', 'utf-8').includes('musl');
    }
    catch (_a) {
        return null;
    }
};
var isMuslFromReport = function () {
    var _a;
    var report = null;
    if (typeof ((_a = process.report) === null || _a === void 0 ? void 0 : _a.getReport) === 'function') {
        process.report.excludeNetwork = true;
        report = process.report.getReport();
    }
    if (!report) {
        return null;
    }
    if (report.header && report.header.glibcVersionRuntime) {
        return false;
    }
    if (Array.isArray(report.sharedObjects)) {
        if (report.sharedObjects.some(isFileMusl)) {
            return true;
        }
    }
    return false;
};
var isMuslFromChildProcess = function () {
    try {
        return require('child_process').execSync('ldd --version', { encoding: 'utf8' }).includes('musl');
    }
    catch (e) {
        return false;
    }
};
function requireNative() {
    var _a, _b, _c, _d;
    if (process.env.NAPI_RS_NATIVE_LIBRARY_PATH) {
        try {
            return require(process.env.NAPI_RS_NATIVE_LIBRARY_PATH);
        }
        catch (err) {
            loadErrors.push(err);
        }
    }
    else if (process.platform === 'android') {
        if (process.arch === 'arm64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-android-arm64');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-android-arm64');
                var bindingPackageVersion = require('yellowstone-grpc-napi-android-arm64/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else if (process.arch === 'arm') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-android-arm-eabi');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-android-arm-eabi');
                var bindingPackageVersion = require('yellowstone-grpc-napi-android-arm-eabi/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else {
            loadErrors.push(new Error("Unsupported architecture on Android ".concat(process.arch)));
        }
    }
    else if (process.platform === 'win32') {
        if (process.arch === 'x64') {
            if (((_b = (_a = process.config) === null || _a === void 0 ? void 0 : _a.variables) === null || _b === void 0 ? void 0 : _b.shlib_suffix) === 'dll.a' || ((_d = (_c = process.config) === null || _c === void 0 ? void 0 : _c.variables) === null || _d === void 0 ? void 0 : _d.node_target_type) === 'shared_library') {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-win32-x64-gnu');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-win32-x64-gnu');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-win32-x64-gnu/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
            else {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-win32-x64-msvc');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-win32-x64-msvc');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-win32-x64-msvc/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
        }
        else if (process.arch === 'ia32') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-win32-ia32-msvc');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-win32-ia32-msvc');
                var bindingPackageVersion = require('yellowstone-grpc-napi-win32-ia32-msvc/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else if (process.arch === 'arm64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-win32-arm64-msvc');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-win32-arm64-msvc');
                var bindingPackageVersion = require('yellowstone-grpc-napi-win32-arm64-msvc/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else {
            loadErrors.push(new Error("Unsupported architecture on Windows: ".concat(process.arch)));
        }
    }
    else if (process.platform === 'darwin') {
        try {
            return require('@triton-one/yellowstone-grpc-napi-darwin-universal');
        }
        catch (e) {
            loadErrors.push(e);
        }
        try {
            var binding = require('yellowstone-grpc-napi-darwin-universal');
            var bindingPackageVersion = require('yellowstone-grpc-napi-darwin-universal/package.json').version;
            if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
            }
            return binding;
        }
        catch (e) {
            loadErrors.push(e);
        }
        if (process.arch === 'x64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-darwin-x64');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-darwin-x64');
                var bindingPackageVersion = require('yellowstone-grpc-napi-darwin-x64/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else if (process.arch === 'arm64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-darwin-arm64');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-darwin-arm64');
                var bindingPackageVersion = require('yellowstone-grpc-napi-darwin-arm64/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else {
            loadErrors.push(new Error("Unsupported architecture on macOS: ".concat(process.arch)));
        }
    }
    else if (process.platform === 'freebsd') {
        if (process.arch === 'x64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-freebsd-x64');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-freebsd-x64');
                var bindingPackageVersion = require('yellowstone-grpc-napi-freebsd-x64/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else if (process.arch === 'arm64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-freebsd-arm64');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-freebsd-arm64');
                var bindingPackageVersion = require('yellowstone-grpc-napi-freebsd-arm64/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else {
            loadErrors.push(new Error("Unsupported architecture on FreeBSD: ".concat(process.arch)));
        }
    }
    else if (process.platform === 'linux') {
        if (process.arch === 'x64') {
            if (isMusl()) {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-x64-musl');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-x64-musl');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-x64-musl/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
            else {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-x64-gnu');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-x64-gnu');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-x64-gnu/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
        }
        else if (process.arch === 'arm64') {
            if (isMusl()) {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-arm64-musl');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-arm64-musl');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-arm64-musl/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
            else {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-arm64-gnu');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-arm64-gnu');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-arm64-gnu/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
        }
        else if (process.arch === 'arm') {
            if (isMusl()) {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-arm-musleabihf');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-arm-musleabihf');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-arm-musleabihf/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
            else {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-arm-gnueabihf');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-arm-gnueabihf');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-arm-gnueabihf/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
        }
        else if (process.arch === 'loong64') {
            if (isMusl()) {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-loong64-musl');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-loong64-musl');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-loong64-musl/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
            else {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-loong64-gnu');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-loong64-gnu');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-loong64-gnu/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
        }
        else if (process.arch === 'riscv64') {
            if (isMusl()) {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-riscv64-musl');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-riscv64-musl');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-riscv64-musl/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
            else {
                try {
                    return require('@triton-one/yellowstone-grpc-napi-linux-riscv64-gnu');
                }
                catch (e) {
                    loadErrors.push(e);
                }
                try {
                    var binding = require('yellowstone-grpc-napi-linux-riscv64-gnu');
                    var bindingPackageVersion = require('yellowstone-grpc-napi-linux-riscv64-gnu/package.json').version;
                    if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                        throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                    }
                    return binding;
                }
                catch (e) {
                    loadErrors.push(e);
                }
            }
        }
        else if (process.arch === 'ppc64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-linux-ppc64-gnu');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-linux-ppc64-gnu');
                var bindingPackageVersion = require('yellowstone-grpc-napi-linux-ppc64-gnu/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else if (process.arch === 's390x') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-linux-s390x-gnu');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-linux-s390x-gnu');
                var bindingPackageVersion = require('yellowstone-grpc-napi-linux-s390x-gnu/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else {
            loadErrors.push(new Error("Unsupported architecture on Linux: ".concat(process.arch)));
        }
    }
    else if (process.platform === 'openharmony') {
        if (process.arch === 'arm64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-openharmony-arm64');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-openharmony-arm64');
                var bindingPackageVersion = require('yellowstone-grpc-napi-openharmony-arm64/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else if (process.arch === 'x64') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-openharmony-x64');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-openharmony-x64');
                var bindingPackageVersion = require('yellowstone-grpc-napi-openharmony-x64/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else if (process.arch === 'arm') {
            try {
                return require('@triton-one/yellowstone-grpc-napi-openharmony-arm');
            }
            catch (e) {
                loadErrors.push(e);
            }
            try {
                var binding = require('yellowstone-grpc-napi-openharmony-arm');
                var bindingPackageVersion = require('yellowstone-grpc-napi-openharmony-arm/package.json').version;
                if (bindingPackageVersion !== '0.2.0' && process.env.NAPI_RS_ENFORCE_VERSION_CHECK && process.env.NAPI_RS_ENFORCE_VERSION_CHECK !== '0') {
                    throw new Error("Native binding package version mismatch, expected 0.2.0 but got ".concat(bindingPackageVersion, ". You can reinstall dependencies to fix this issue."));
                }
                return binding;
            }
            catch (e) {
                loadErrors.push(e);
            }
        }
        else {
            loadErrors.push(new Error("Unsupported architecture on OpenHarmony: ".concat(process.arch)));
        }
    }
    else {
        loadErrors.push(new Error("Unsupported OS: ".concat(process.platform, ", architecture: ").concat(process.arch)));
    }
}
nativeBinding = requireNative();
if (!nativeBinding || process.env.NAPI_RS_FORCE_WASI) {
    var wasiBinding = null;
    var wasiBindingError = null;
    try {
        wasiBinding = require('./yellowstone-grpc-napi.wasi.cjs');
        nativeBinding = wasiBinding;
    }
    catch (err) {
        if (process.env.NAPI_RS_FORCE_WASI) {
            wasiBindingError = err;
        }
    }
    if (!nativeBinding || process.env.NAPI_RS_FORCE_WASI) {
        try {
            wasiBinding = require('yellowstone-grpc-napi-wasm32-wasi');
            nativeBinding = wasiBinding;
        }
        catch (err) {
            if (process.env.NAPI_RS_FORCE_WASI) {
                if (!wasiBindingError) {
                    wasiBindingError = err;
                }
                else {
                    wasiBindingError.cause = err;
                }
                loadErrors.push(err);
            }
        }
    }
    if (process.env.NAPI_RS_FORCE_WASI === 'error' && !wasiBinding) {
        var error = new Error('WASI binding not found and NAPI_RS_FORCE_WASI is set to error');
        error.cause = wasiBindingError;
        throw error;
    }
}
if (!nativeBinding) {
    if (loadErrors.length > 0) {
        throw new Error("Cannot find native binding. " +
            "npm has a bug related to optional dependencies (https://github.com/npm/cli/issues/4828). " +
            'Please try `npm i` again after removing both package-lock.json and node_modules directory.', {
            cause: loadErrors.reduce(function (err, cur) {
                cur.cause = err;
                return cur;
            }),
        });
    }
    throw new Error("Failed to load native binding");
}
module.exports = nativeBinding;
module.exports.DuplexStream = nativeBinding.DuplexStream;
module.exports.GrpcClient = nativeBinding.GrpcClient;
module.exports.decodeTxError = nativeBinding.decodeTxError;
module.exports.encodeTx = nativeBinding.encodeTx;
module.exports.JsCompressionAlgorithm = nativeBinding.JsCompressionAlgorithm;
module.exports.WasmUiTransactionEncoding = nativeBinding.WasmUiTransactionEncoding;
